dll/win32/msgina/gui.c

```diff
@@ -57,57 +64,219 @@ DlgData_Create(HWND hwndDlg, PGINA_CONTEXT pgContext)
+    xNew = xOld - (cxNew - cxOld) / 2;
+    yNew = yOld - (cyNew - cyOld) / 2;
+    MoveWindow(hwndDlg, xNew, yNew, cxNew, cyNew, TRUE);
+}
+
+/*static*/ VOID
+DlgData_MoveControls(
+    _In_ HWND hwndDlg,
+    _In_opt_ PDLG_DATA pDlgData)
+{
+    HWND hwndLogo, hwndBar;
+    UINT nNumChildren = 0;
+    HWND hWnd;
+    HDWP hdwp;
+    RECT rc, rcCtrl;
 
-    pDlgData = (PDLG_DATA)GetWindowLongPtrW(hwndDlg, GWLP_USERDATA);
     if (!pDlgData)
     {
+        pDlgData = (PDLG_DATA)GetWindowLongPtrW(hwndDlg, GWLP_USERDATA);
+        if (!pDlgData)
+            return;
+    }
+
+    // TODO: Decide whether to "horizontally-center" the bitmaps wrt. the dialog.
+
+#define _ID2STR(id) L"#" #id
+#define ID2STR(id)  _ID2STR(id)
+
+    /* Create and place the logo control */
+    hwndLogo = CreateWindowExW(0,
+                               WC_STATICW,
+                               ID2STR(IDI_ROSLOGO),
+                               WS_VISIBLE | WS_CHILD | SS_BITMAP,
+                               0, 0,
+                               pDlgData->LogoWidth,
+                               pDlgData->LogoHeight,
+                               hwndDlg,
+                               (HMENU)IDC_ROSLOGO,
+                               hDllInstance,
+                               NULL);
+
+    /* If the logo failed to be created, don't bother creating the bar and moving anything */
+    if (!hwndLogo)
         return;
+
+    /* Create and place the bar control just beneath the logo */
+    hwndBar = CreateWindowExW(0,
+                              WC_STATICW,
+                              ID2STR(IDI_BAR),
+                              WS_VISIBLE | WS_CHILD | SS_BITMAP /*| SS_OWNERDRAW*/,
+                              0, pDlgData->LogoHeight,
+                              pDlgData->BarWidth, // LogoWidth
+                              pDlgData->BarHeight,
+                              hwndDlg,
+                              (HMENU)IDC_BAR,
+                              hDllInstance,
+                              NULL);
+
+#undef ID2STR
+
+    /* Retrieve the total height of logo + bar */
+    SetRectEmpty(&rcCtrl);
+#if 0
+    GetWindowRect(hwndLogo, &rc);
+    MapWindowPoints(NULL, hwndDlg, (LPPOINT)&rc, sizeof(RECT)/sizeof(POINT));
+    rcCtrl.bottom += rc.bottom - rc.top;
+
+    if (hwndBar)
+    {
+        GetWindowRect(hwndBar, &rc);
+        MapWindowPoints(NULL, hwndDlg, (LPPOINT)&rc, sizeof(RECT)/sizeof(POINT));
+        rcCtrl.bottom += rc.bottom - rc.top;
+    }
+#else
+    rcCtrl.bottom += pDlgData->LogoHeight;
+    if (hwndBar)
+        rcCtrl.bottom += pDlgData->BarHeight;
+#endif
+
+    /* Count the number of children windows in this dialog */
+    for (hWnd = GetWindow(hwndDlg, GW_CHILD);
+         hWnd;
+         hWnd = GetNextWindow(hWnd, GW_HWNDNEXT))
+    {
+        /* Skip the logo and bar controls */
+        // LONG_PTR nID = GetWindowLongPtrW(hWnd, GWLP_ID);
+        // if (nID == IDC_ROSLOGO || nID == IDC_BAR)
```

Comment:
Why keeping those 2 commented lines?

Reply:
Notice how the PR is in *draft* state. This means that there are a lot of unfinished/untested/in-progress stuff ...